dev-support/hbase-personality.sh

```diff
@@ -816,6 +816,69 @@ function hbaseanti_patchfile
   return 0
 }
 
+######################################
+
+add_test_type visible-for-testing
+
+## @description  visible-for-testing file filter
+## @audience     private
+## @stability    evolving
+## @param        filename
+function visible-for-testing_filefilter
+{
+  local filename=$1
+
+  if [[ ${filename} =~ \.java$ ]]; then
+    add_test visible-for-testing
+  fi
+}
+
+## @description  visible-for-testing patch file check
+## @audience     private
+## @stability    evolving
+## @param        repostatus
+function visible-for-testing_precompile
+{
+  local repostatus=$1
+  local logfile="${PATCH_DIR}/patch-visible-for-testing.txt"
+  local errors
+
+  if [[ "${repostatus}" = branch ]]; then
+    return 0
+  fi
+
+  if ! verify_needed_test visible-for-testing; then
+    return 0
+  fi
+
+  big_console_header "VisibleForTesting plugin: ${BUILDMODE}"
+  if [[ "${BUILDMODE}" == "full" ]]; then
+    yetus_debug "going to check the whole repo"
+    find -type f -name *.java | grep -v generated | xargs grep -l "@InterfaceAudience.Public\|@InterfaceAudience.LimitedPrivate" | xargs grep -l @VisibleForTesting >> ${logfile}
```

Comment:
I would prefer we reject VisibleForTesting globally. 

What purpose does it serve? An annotation is useful when it supports tooling, like an IDE, or some static analysis (error-prone?). We don't have anything like that depending on VisibleForTesting. So in my view VisibleForTesting is pointless and, worse, brings in Guava, which has a history of problematic compatibility problems. What happens when Google decides to remove VisibleForTesting from Guava some day? Why should we set ourselves up for that risk. There must be some reason. I see no good reason.

Reply:
If we are in a company I think I will agree with you to set a strict rule to prevent any potential risks. But in a open source community, I think the starting point is a bit different. I agree with you that, a VisibleForTesting annotation is like a javadoc, but I do not think this is strong enough reason to disable one way, we could accept both ways, to suit more developers in the world. As you point out Guava has a history to introduce conflicts, but our solution is to shade it and purge it from the public API so it will not effect downstream users, but we still want to use it instead of disable it right? It is widely used across the java developers.

FWIW, we have not seen a plan of google to remove this class. It is also used in Android

https://developer.android.com/reference/androidx/annotation/VisibleForTesting

And it is not very difficult to just purge it from the source code in the future if we have more strong reasons to disable it, just like what you have done. You can use semantice search in IDE instead of text search so you could find all the reference quickly and accurately.

So I prefer that, we have tool to disable the usage of this annotation in IA.Public and IA.LimitedPrivate classes, as this is the actual anti-pattern here. And in the future, if we find out that VisibleForTesting also introduces problems even in IA.Private classes, we could just ban it in maven enforcer.

What do you think?

Thanks.